### DISCUSSION TITLE
Problem creating docker container with v22.10

### DISCUSSION THREAD

**Discussion by franz-gassner** [nextflow-io/nextflow#3357] [Q&A] [answered]
## Bug report 

I recently upgraded nextflow from v21.04.3.5560 to v22.10.0.5826 and encountered a problem involving creation of a docker container.

### Expected behavior and actual behavior

Here is the docker section in the nextflow.config file:

```
docker {
    enabled = true
    docker.runOptions='-u $(id -u):$(id -g) -v /home/m3fga/limcr-ngs/databases/hg38/:/home/m3fga/limcr-ngs/databases/hg38/'
}
```
And here the failing process in the .nf file:

```
process runBamReadcount {
    tag { "${params.projectName}.rBamReadcount" }
    publishDir "${params.varscan_outDir}", mode: 'copy', pattern: "*.readcounts"
    container 'mgibio/bam-readcount:latest'
    
    input:
        tuple val(sample), file(snp_filter_reg), file(indel_reg), file(cll_bam), file(cll_bai)
    
    output:
        tuple val(sample), file("*.snp.Somatic.hc.filter.readcounts"), file("*.indel.Somatic.hc.readcounts")
        
    """
    bam-readcount \
      -q 1 -b 20 -w 1 \
      -f ${params.hg38_ref} \
      -l ${snp_filter_reg} \
      ${cll_bam} \
      > "${sample}".VarScan2.4.4.snp.Somatic.hc.filter.readcounts
    
    bam-readcount \
      -q 1 -b 20 -w 1 \
      -f ${params.hg38_ref} \
      -l ${indel_reg} \
      ${cll_bam} \
      > "${sample}".VarScan2.4.4.indel.Somatic.hc.readcounts
    """
}
```
### Steps to reproduce the problem

I checked all involved files, they are ok. When using Nextflow v21.04.3.5560, no error occured. I also created the docker container manually (i.e. without nextflow) and used the same files as input, and the analysis also worked as intended. There was no alteration in the pipeline except upgrade of Nextflow from v21.04.3.5560 to v22.10.0.5826.

### Program output 

This is the error report of the failed process copied from terminal:

```
Error executing process > 'runBamReadcount (Oncotyrol_DNA_reanalysis.rBamReadcount)'

Caused by:
  Process `runBamReadcount (Oncotyrol_DNA_reanalysis.rBamReadcount)` terminated with an error exit status (1)

Command executed:

  bam-readcount       -q 1 -b 20 -w 1       -f /home/m3fga/limcr-ngs/databases/hg38/hg38.fa       -l 0106.VarScan2.4.4.snp.Somatic.hc.filter.regions       0106.CLL.bwa.sorted.calmd.marked.realigned.recalBam.bam       > "0106".VarScan2.4.4.snp.Somatic.hc.filter.readcounts

  bam-readcount       -q 1 -b 20 -w 1       -f /home/m3fga/limcr-ngs/databases/hg38/hg38.fa       -l 0106.VarScan2.4.4.indel.Somatic.hc.regions       0106.CLL.bwa.sorted.calmd.marked.realigned.recalBam.bam       > "0106".VarScan2.4.4.indel.Somatic.hc.readcounts

Command exit status:
  1

Command output:
  (empty)

Command error:
  Minimum mapping quality is set to 0
  [E::hts_hopen] Failed to open file /bin/bash
  [E::hts_open_format] Failed to open file "/bin/bash" : Exec format error
  Fail to open BAM file /bin/bash
```

### Environment 

* Nextflow version: 22.10.0.5826
* Java version: "11.0.14" 2022-01-18 LTS
* Operating system: Linux Red Hat Enterprise Linux 8.5
* Bash version: GNU bash, version 4.4.20(1)-release (x86_64-redhat-linux-gnu)

### Additional context

Finally I checked the .command.run files of both runs:

Nextflow v21.04.3.5560
line 253:
```
nxf_launch() {
    docker run -i --cpus 1.0 -e "NXF_DEBUG=${NXF_DEBUG:=0}" -v /limcr-ngs/Franz/2021-09-27_Nextflow_Test/work:/limcr-ngs/Franz/2021-09-27_Nextflow_Test/work -v "$PWD":"$PWD" -w "$PWD" --entrypoint /bin/bash -u $(id -u):$(id -g) -v /home/m3fga/limcr-ngs/databases/hg19/:/home/m3fga/limcr-ngs/databases/hg19/ --name $NXF_BOXID mgibio/bam-readcount:latest -c "/bin/bash /limcr-ngs/Franz/2021-09-27_Nextflow_Test/work/48/964fb6b5904e08b2c83de8cf48eeff/.command.run nxf_trace"
}
```

Nextflow v22.10.0.5826
line 274
```
nxf_launch() {
    docker run -i --cpus 1.0 -e "NXF_DEBUG=${NXF_DEBUG:=0}" -v /limcr-ngs/Franz/2022-10-19_Oncotyrol_DNA_reanalysis_nextflow/work:/limcr-ngs/Franz/2022-10-19_Oncotyrol_DNA_reanalysis_nextflow/work -w "$PWD" -u $(id -u):$(id -g) -v /home/m3fga/limcr-ngs/databases/hg38/:/home/m3fga/limcr-ngs/databases/hg38/ --name $NXF_BOXID mgibio/bam-readcount:latest /bin/bash /limcr-ngs/Franz/2022-10-19_Oncotyrol_DNA_reanalysis_nextflow/work/b2/c40a168828e421ab2b5d97aa43a148/.command.run nxf_trace
}
```

I suspect the root of the error here. In the second run, bam-readcount is given "/bin/bash/" as argument, while it expects a bam file. Furthermore, the "-c" argument is missing in the second run.

Can you maybe look at this issue and help me with that?

Thanks a lot!

Best,
Franz

**Comment by pditommaso** [comment]
This happens because the container uses a custom entry point and as of version 22.08.0-edge nextflow does not override it anymore with `/bin/bash` 

https://github.com/nextflow-io/nextflow/blob/master/changelog.txt#L181


You have to make sure your container uses a shell as entry point. In alternative you can force nextflow using the old approach setting the following variable in your env 

```
export NXF_CONTAINER_ENTRYPOINT_OVERRIDE=true
```

**Comment by franz-gassner** [comment, marked as answer]
Following your suggestion, I added the entry point argument to the docker call in the nextflow.config file:

`docker.runOptions='-u $(id -u):$(id -g) --entrypoint="" -v /home/m3fga/limcr-ngs/databases/hg38/:/home/m3fga/limcr-ngs/databases/hg38/'`

This resolved the issue.

Thanks a lot for your quick help!

**Reply by pditommaso** [reply]
That's a nice trick 👍
